### PR TITLE
Fix and improve path finding

### PR DESCRIPTION
**Pull request by bartekn**
This PR fixes several bugs and adds improvements making path finding results better, returning them in a shorter time.

The most severe bug was found in `pathNode.Cost` method. `pathNode` is a linked list node with `Tail` field pointing to the next asset on the path. So the front of the list is the currently extended source asset and the tail is always the destination asset. The previous code was wrongly consuming offers in the wrong order. Fixing this bug required to walk through the current path from the tail (the destination asset) to front (the current source).

I added a new [test](https://github.com/stellar/go/pull/719/files#diff-dd055dd31266cf1b960386b17e35a83aR121) explaining why the old code was wrong. Consider the path: `AAA -> BBB -> CCC`. The core of the problem is that the previous code was using `DestinationAmount` as a starting value for `AAA -> BBB` trade, in other words: "give me amount of `AAA` needed to buy `DestinationAmount` of `BBB`". It doesn't make any sense. What you should do first is: "give me amount of `BBB` needed to buy `DestinationAmount` of `CCC`". So for the destination amount equal `10` and the following offers (`scc` notation):
```
offer :trader, {for:["AAA", :gateway], sell:["BBB", :gateway]}, 1, 11
offer :trader, {for:["BBB", :gateway], sell:["CCC", :gateway]}, 10, 0.1
```
path finding trade order 1

The old algorithm would not find a path as it is not possible to buy `1 BBB` with `10 AAA` and `1 BBB` is needed to buy `10 CCC`.

Fixing this introduced a huge optimization opportunity. Basically the cost of trades between current source asset and the destination asset does not change after prepending the path with a new asset. So we can cache the results instead of checking every order book on the path every time we add a new asset in the front of the path.

Another improvement was to remove `A -> B -> A` subpaths from the paths as they will always be a bad deal. This also removed all `destination asset -> A -> destination asset` paths.

Finally, added `MaxPathLength` that is configurable via CLI switch or environment variable and defines the maximum length of the path that `/paths` will return, with a default value of `4` (`source asset -> A -> B -> destination asset`). The default value of 4 comes as a result of observation of paths found between many different assets: most of them have the maximum length of 4 and if the path length was higher it was a bad deal. The `MaxPathLength` allowed another optimization in `search.extendSearch` method. Basically if `p.Depth == s.MaxLength-1` and connected asset is not a target we don't extend such path as it will not be possible to extend the path any longer.

After implementing all the optimizations average search time for my queries dropped from 10 seconds to 2 seconds giving a better results.

Finally, a single search is run in `REPEATABLE READ, READ ONLY` DB transaction to have a stable view of the offers table.

### CODE REVIEW

File: services/horizon/internal/actions/helpers.go [hunk 1]
```diff
@@ -237,7 +237,7 @@ func (base *Base) GetAccountID(name string) (result xdr.AccountId) {
 // conventions
 func (base *Base) GetAmount(name string) (result xdr.Int64) {
 	var err error
-	result, err = amount.Parse(base.GetString("destination_amount"))
```
Comment (nikhilsaraf) [inline services/horizon/internal/actions/helpers.go]: why did this change -- was the name param never used correctly earlier?

File: services/horizon/internal/simplepath/path.go [hunk 2]
```diff
@@ -68,29 +87,56 @@ func (p *pathNode) Cost(amount xdr.Int64) (xdr.Int64, error) {
...
+	}
+
+	// The first element of the current path is the current source asset.
+	// The last element (with `Tail` = nil) of the current path is the destination
+	// asset. To make the calculations correct we start by selling destination
+	// asset to the second from the end asset and continue until we reach the current
+	// source asset.
 	cur := p
+	stack := make([]*pathNode, 0, p.Depth)
 	for cur.Tail != nil {
```
Comment (nikhilsaraf) [inline services/horizon/internal/simplepath/path.go]: I think pathNode (cur) here may be inserted backwards (i.e. destination asset first and source asset last) -- see the insertion order of computedNode in search.go#Init()

File: services/horizon/internal/actions/helpers.go (hunk 1 above)
Comment (bartekn) [reply services/horizon/internal/actions/helpers.go]: Yes, this was working correctly only when `name="destination_amount`.

**Comment by bartekn** [comment]
Thanks @nikhilsaraf! I will merge it now and it will be autodeployed to https://horizon-dev-pubnet.stellar.org/ so more people can easily test it. In the end we can revert this PR is there's a need but I believe it should work fine and give better results.

**Pull request merged**